internal/controller/dashboard_controller.go

```diff
@@ -147,162 +145,41 @@ func (r *DashboardReconciler) SetupWithManager(mgr ctrl.Manager) error {
 // reconcileCreate ensures the Grafana dashboard described in configmap is created in Grafana.
 // This function is also responsible for:
 // - Adding the finalizer to the configmap
-func (r DashboardReconciler) reconcileCreate(ctx context.Context, grafanaAPI *grafanaAPI.GrafanaHTTPAPI, dashboard *v1.ConfigMap) (ctrl.Result, error) { // nolint:unparam
+func (r DashboardReconciler) reconcileCreate(ctx context.Context, grafanaAPI *grafanaAPI.GrafanaHTTPAPI, dashboard *v1.ConfigMap) error { // nolint:unparam
 	// Add finalizer first if not set to avoid the race condition between init and delete.
 	finalizerAdded, err := r.finalizerHelper.EnsureAdded(ctx, dashboard)
 	if err != nil || finalizerAdded {
-		return ctrl.Result{}, errors.WithStack(err)
-	}
-
-	// Configure the dashboard in Grafana
-	if err := r.configureDashboard(ctx, grafanaAPI, dashboard); err != nil {
-		return ctrl.Result{}, errors.WithStack(err)
-	}
-
-	return ctrl.Result{}, nil
-}
-
-func getDashboardUID(dashboard map[string]interface{}) (string, error) {
-	UID, ok := dashboard["uid"].(string)
-	if !ok {
-		return "", errors.New("dashboard UID not found in configmap")
-	}
-	return UID, nil
-}
-
-func cleanDashboardID(dashboard map[string]interface{}) {
-	if dashboard["id"] != nil {
-		delete(dashboard, "id")
-	}
-}
-
-func getOrgFromDashboardConfigmap(dashboard *v1.ConfigMap) (string, error) {
-	// Try to look for an annotation first
-	annotations := dashboard.GetAnnotations()
-	if annotations != nil && annotations[grafanaOrganizationLabel] != "" {
-		return annotations[grafanaOrganizationLabel], nil
-	}
-
-	// Then look for a label
-	labels := dashboard.GetLabels()
-	if labels != nil && labels[grafanaOrganizationLabel] != "" {
-		return labels[grafanaOrganizationLabel], nil
-	}
-
-	// Return an error if no label was found
-	return "", errors.New("No organization label found in configmap")
-}
-
-func (r DashboardReconciler) configureDashboard(ctx context.Context, grafanaAPI *grafanaAPI.GrafanaHTTPAPI, dashboardCM *v1.ConfigMap) error {
-	logger := log.FromContext(ctx)
-
-	dashboardOrg, err := getOrgFromDashboardConfigmap(dashboardCM)
-	if err != nil {
-		logger.Error(err, "Skipping dashboard, no organization found")
-		return nil
+		return errors.WithStack(err)
 	}
 
-	// TODO Tenant Governance: Filter the dashboards with the list of authorized tenants
+	service := grafana.NewService(r.Client, grafanaAPI)
```

Comment:
Why not create the service earlier (before the calls to reconcileCreate and delete)?

Reply:
Good point, I moved this to earlier stage in Reconcile function.